? lib/winime
? programs/Xserver/hw/xwin/IMdkit
? programs/Xserver/hw/xwin/winic.c
? programs/Xserver/hw/xwin/winic.h
? programs/Xserver/hw/xwin/winime.c
? programs/Xserver/hw/xwin/winimserver.c
Index: config/cf/X11.tmpl
===================================================================
RCS file: /cvs/xorg/xc/config/cf/X11.tmpl,v
retrieving revision 1.38
diff -u -r1.38 X11.tmpl
--- config/cf/X11.tmpl	3 Dec 2004 03:33:25 -0000	1.38
+++ config/cf/X11.tmpl	16 Jan 2005 09:19:43 -0000
@@ -708,6 +708,10 @@
 #define BuildWindowsWMLibrary	NO
 #endif
 
+#ifndef BuildWinIMELibrary
+#define BuildWinIMELibrary	NO
+#endif
+
 #ifndef BuildMiscDocs
 #define BuildMiscDocs		NO
 #endif
@@ -2364,6 +2368,30 @@
 #define ProfileLibWindowsWM	NO
 #endif
 
+#if BuildWinIMELibrary
+#ifndef SharedLibWinIME
+#define SharedLibWinIME	HasSharedLibraries
+#endif
+#ifndef NormalLibWinIME
+#define NormalLibWinIME	(!SharedLibWinIME || ForceNormalLib)
+#endif
+#ifndef DebugLibWinIME
+#define DebugLibWinIME		NO
+#endif
+#ifndef ProfileLibWinIME
+#define ProfileLibWinIME	NO
+#endif
+#else
+#undef  SharedLibWinIME
+#define SharedLibWinIME		NO
+#undef  NormalLibWinIME
+#define NormalLibWinIME		NO
+#undef  DebugLibWinIME
+#define DebugLibWinIME		NO
+#undef  ProfileLibWinIME
+#define ProfileLibWinIME	NO
+#endif
+
 #if BuildGLULibrary
 #ifndef SharedLibGlu
 #define SharedLibGlu		HasSharedLibraries
@@ -2766,6 +2794,16 @@
 ProjectUnsharedLibReferences(WINDOWSWM,WindowsWM,$(WINDOWSWMLIBSRC),XBuildLibDir)
 #endif
 
+   WINIMELIBSRC = $(LIBSRC)/winime
+#if SharedLibWinIME
+#ifndef SharedWinIMERev
+#define SharedWinIMERev 1.0
+#endif
+SharedLibReferences(WINIME,WinIME,$(WINIMELIBSRC),SOWINIMEREV,SharedWinIMERev)
+#else
+ProjectUnsharedLibReferences(WINIME,WinIME,$(WINIMELIBSRC),XBuildLibDir)
+#endif
+
 # ifndef SharedLibXfontcache
 #  define SharedLibXfontcache	HasSharedLibraries
 # endif
Index: config/cf/cygwin.cf
===================================================================
RCS file: /cvs/xorg/xc/config/cf/cygwin.cf,v
retrieving revision 1.17
diff -u -r1.17 cygwin.cf
--- config/cf/cygwin.cf	15 Nov 2004 15:06:52 -0000	1.17
+++ config/cf/cygwin.cf	16 Jan 2005 09:19:43 -0000
@@ -314,6 +314,10 @@
 # define BuildWindowsWMLibrary	NO
 #endif /* BuildXWinMultiWindowExtWM && BuildWindowsWMLibrary */
 
+#if !defined(BuildWinIMELibrary)
+# define BuildWinIMELibrary	YES
+#endif
+
 /* XWin Server specific defines */
 #if BuildXWinClipboard
 # define XWinClipboardDefines	-DXWIN_CLIPBOARD
@@ -356,6 +360,11 @@
 #else
 # define XWinXF86ConfigDefines 
 #endif /* BuildXWinXF86Config */
+#if BuildWinIMELibrary
+# define XWinWinIMEDefines	-DXWIN_WINIME
+#else
+# define XWinWinIMEDefines 
+#endif /* BuildXWinIMELibrary */
 
 /*
  * XFree86Server is defined for the w32api headers, which protects some
@@ -371,6 +380,7 @@
 				XWinUpdateStatsDefines \
 				XWinClipboardDefines XWinMultiWindowDefines \
 				XWinMultiWindowExtWMDefines \
+				XWinWinIMEDefines \
 				-DDDXBEFORERESET
 #define ServerOSDefines		-DDDXTIME -DDDXOSINIT \
 				-DDDXOSVERRORF -DDDXOSFATALERROR
Index: config/cf/cygwin.rules
===================================================================
RCS file: /cvs/xorg/xc/config/cf/cygwin.rules,v
retrieving revision 1.9
diff -u -r1.9 cygwin.rules
--- config/cf/cygwin.rules	2 Sep 2004 01:10:28 -0000	1.9
+++ config/cf/cygwin.rules	16 Jan 2005 09:19:43 -0000
@@ -31,6 +31,7 @@
 #define SharedLibDpsTk		YES
 #define SharedLibGlu		YES
 #define SharedLibWindowsWM	NO
+#define SharedLibWinIME		NO
 #ifndef SharedDataSeparation
 #define SharedDataSeparation	NO
 #endif
Index: config/cf/mingw.cf
===================================================================
RCS file: /cvs/xorg/xc/config/cf/mingw.cf,v
retrieving revision 1.4
diff -u -r1.4 mingw.cf
--- config/cf/mingw.cf	10 Jan 2005 10:48:46 -0000	1.4
+++ config/cf/mingw.cf	16 Jan 2005 09:19:44 -0000
@@ -301,6 +301,10 @@
 # define BuildWindowsWMLibrary	NO
 #endif /* BuildXWinMultiWindowExtWM && BuildWindowsWMLibrary */
 
+#if !defined(BuildWinIMELibrary)
+# define BuildWinIMELibrary	YES
+#endif
+
 #if (BuildXWinClipboard || BuildXWinMultiWindow || BuildXWinMultiWindowExtWM) && !defined(BuildX11Lib)
 #  define BuildX11Lib YES
 #  define NormalLibX11 YES
Index: lib/Imakefile
===================================================================
RCS file: /cvs/xorg/xc/lib/Imakefile,v
retrieving revision 1.11
diff -u -r1.11 Imakefile
--- lib/Imakefile	29 Sep 2004 12:40:28 -0000	1.11
+++ lib/Imakefile	16 Jan 2005 09:20:02 -0000
@@ -161,6 +161,10 @@
 WINDOWSLIBDIR = windows
 #endif
 
+#if BuildWinIMELibrary
+WINIMELIBDIR = winime
+#endif
+
 XF86EXTLIBS = $(XF86MISCLIBDIR) $(XF86VMLIBDIR) \
 		$(XF86DGALIBDIR) $(XF86RUSHLIBDIR)
 
@@ -236,7 +240,8 @@
              $(FONTCONFIGBUILDDIR) $(XFT1LIBDIR) \
 	     $(XFTLIBDIR) $(XVMCLIBDIR) $(RANDRLIBDIR) $(XTRAPLIBDIR) \
 	     $(XRESLIBDIR) $(XCURSORLIBDIR) $(APPLELIBDIR) $(DMXLIBDIR) \
-	     $(WINDOWSLIBDIR) $(XEVIELIBDIR) $(XFIXESLIBDIR) $(DAMAGELIBDIR) $(COMPOSITELIBDIR)
+	     $(WINDOWSLIBDIR) $(XEVIELIBDIR) $(XFIXESLIBDIR) $(DAMAGELIBDIR) \
+	     $(COMPOSITELIBDIR) $(WINIMELIBDIR)
 
 SUBDIRS = $(BERKDIR) xtrans $(LINTSUBDIRS) $(FONTSUBDIR) $(FONTENCSUBDIR) \
           $(FONTCACHELIBDIR)
Index: programs/Xserver/Imakefile
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/Imakefile,v
retrieving revision 1.26
diff -u -r1.26 Imakefile
--- programs/Xserver/Imakefile	9 Jan 2005 20:45:14 -0000	1.26
+++ programs/Xserver/Imakefile	16 Jan 2005 09:20:16 -0000
@@ -1029,13 +1029,29 @@
 #else
 XWINOPENGLLIB = 
 #endif
+#if BuildWinIMELibrary
+XWINIMMLIB = -limm32
+#else
+XWINIMMLIB = 
+#endif
+#if BuildWinIMELibrary
+XWINIMDKITDIR = $(XWINDDXDIR)/IMdkit
+XWINIMDKITLIB = $(XWINDDXDIR)/IMdkit/LibraryTargetName(Ximd)
+XWINWINIMELIB = -lWinIME -lXext
+#else
+XWINIMDKITDIR =
+XWINIMDKITLIB =
+XWINWINIMELIB =
+#endif
 
 XWINLIB = $(XWINDDXDIR)/LibraryTargetName(XWin)
 XWINDIRS = $(STDDIRS) $(FBDIR) $(SHADOWDIR) $(LAYERDIR) $(XWINDDXDIR) \
-	   $(DEPDIRS) $(XWINPARSERDIR) $(ROOTLESSDIR) $(MIDAMAGEDIR)
+	   $(DEPDIRS) $(XWINPARSERDIR) $(ROOTLESSDIR) $(MIDAMAGEDIR) \
+	   $(XWINIMDKITDIR)
 XWINOBJS = $(XWINDDXDIR)/stubs.o $(XWINDDXDIR)/XWin.res
 XWINLIBS = PreFbLibs $(XWINLIB) FbPostFbLibs $(XWINLIB) $(XWINLAYERLIB) \
-           $(SHADOW) $(XWINPARSERLIB) $(ROOTLESSLIB) $(OS)
+           $(SHADOW) $(XWINPARSERLIB) $(ROOTLESSLIB) $(OS) $(XWINIMMLIB) \
+           $(XWINIMDKITLIB)
 #if BuildXWinMultiWindow || BuildXWinClipboard
 XWINX11  = $(XONLYLIB)	   
 # if defined(Win32Architecture)
@@ -1050,16 +1066,19 @@
 #else
 XWINW32  = -lgdi32
 #endif
-XWINSYSLIBS = $(FONTLIBS) $(LDPRELIBS) $(XWINX11) $(SYSLIBS) $(XWINW32)
+XWINSYSLIBS = $(FONTLIBS) $(LDPRELIBS) $(XWINWINIMELIB) $(XWINX11) \
+	      $(SYSLIBS) $(XWINW32) $(XWINIMDKITLIB)
 
 EXTRA_LDOPTIONS = -e _mainCRTStartup
 
 #if HasParallelMake
-MakeMutex($(XWINDIRS) $(XWINOBJS) $(XWINLIB) $(XWINLIBS) $(XWINSYSLIBS))
+MakeMutex($(XWINDIRS) $(XWINOBJS) $(XWINLIB) $(XWINLIBS) $(XWINSYSLIBS) \
+          $(XWINIMDKITLIB))
 #endif
 
 #if ForceServerRemake
-$(XWINOBJS) $(XWINLIB) $(XWINLIBS) $(XWINSYSLIBS):: $(XWINDIRS)
+$(XWINOBJS) $(XWINLIB) $(XWINLIBS) $(XWINSYSLIBS):: \
+ $(XWINDIRS)
 	@if [ -f $@ ]; then touch $@ >/dev/null 2>&1 || exit 0; fi
 #endif
 
Index: programs/Xserver/hw/xwin/Imakefile
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/Imakefile,v
retrieving revision 1.8
diff -u -r1.8 Imakefile
--- programs/Xserver/hw/xwin/Imakefile	6 Jan 2005 16:02:47 -0000	1.8
+++ programs/Xserver/hw/xwin/Imakefile	16 Jan 2005 09:20:34 -0000
@@ -1,6 +1,7 @@
 XCOMM $XFree86: xc/programs/Xserver/hw/xwin/Imakefile,v 1.18 2003/10/02 13:30:09 eich Exp $
 
 #include <Server.tmpl>
+#define IHaveSubdirs
 
 EXTRADEFINES = 
 
@@ -35,6 +36,10 @@
 RC_PROJECT_DEFINES = -DPROJECT_NAME="\\\"$(PROJECT_NAME)\\\""
 EXTRA_DEFINES = $(PROJECT_DEFINES)
 
+#if BuildWinIMELibrary
+SUBDIRS = IMdkit
+#endif
+
 #if BuildXWinClipboard
 SRCS_CLIPBOARD = \
 	winclipboardinit.c \
@@ -67,6 +72,13 @@
 	winwindowswm.c
 #endif
 
+#if BuildWinIMELibrary
+SRCS_WINIMEWM = \
+	winime.c \
+	winic.c \
+	winimserver.c
+#endif
+
 #if BuildXWinNativeGDI
 SRCS_NATIVEGDI = \
 	winclip.c \
@@ -137,6 +149,7 @@
 	$(SRCS_GLX_WINDOWS) \
 	$(SRCS_MULTIWINDOW) \
 	$(SRCS_MULTIWINDOWEXTWM) \
+	$(SRCS_WINIME) \
 	$(SRCS_NATIVEGDI) \
 	$(SRCS_PRIMARYFB) \
 	$(SRCS_RANDR) \
@@ -187,6 +200,13 @@
 	winwindowswm.o
 #endif
 
+#if BuildWinIMELibrary
+OBJS_WINIME = \
+	winime.o \
+	winic.o \
+	winimserver.o
+#endif
+
 #if BuildXWinNativeGDI
 OBJS_NATIVEGDI = \
 	winclip.o \
@@ -257,6 +277,7 @@
 	$(OBJS_GLX_WINDOWS) \
 	$(OBJS_MULTIWINDOW) \
 	$(OBJS_MULTIWINDOWEXTWM) \
+	$(OBJS_WINIME) \
 	$(OBJS_NATIVEGDI) \
 	$(OBJS_PRIMARYFB) \
 	$(OBJS_RANDR) \
@@ -270,7 +291,7 @@
 	   -I$(SERVERSRC)/include -I$(SERVERSRC)/os  \
            -I$(EXTINCSRC) -I$(XINCLUDESRC) \
 	   -I$(SERVERSRC)/render -I$(SERVERSRC)/randr \
-	   -I$(WINDOWSWMLIBSRC)
+	   -I$(WINDOWSWMLIBSRC) -I$(WINIMELIBSRC)
 
 #ifdef XVendorString
 VENDORSTRING = XVendorString
@@ -308,3 +329,6 @@
 InstallManPage(XWinrc,$(MANDIR))
 
 DependTarget()
+
+MakeSubdirs($(SUBDIRS))
+DependSubdirs($(SUBDIRS))
Index: programs/Xserver/hw/xwin/win.h
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/win.h,v
retrieving revision 1.12
diff -u -r1.12 win.h
--- programs/Xserver/hw/xwin/win.h	8 Dec 2004 15:48:15 -0000	1.12
+++ programs/Xserver/hw/xwin/win.h	16 Jan 2005 09:20:34 -0000
@@ -1310,6 +1310,9 @@
 winWindowProc (HWND hWnd, UINT message, 
 	       WPARAM wParam, LPARAM lParam);
 
+void
+winProcessMessage (LPMSG lpMsg);
+
 
 #ifdef XWIN_MULTIWINDOWEXTWM
 /*
@@ -1441,6 +1444,34 @@
 Bool
 winInitCursor (ScreenPtr pScreen);
 
+#ifdef XWIN_WINIME
+/*
+ * winime.c
+ */
+
+void
+winWinIMEExtensionInit ();
+
+void
+winWinIMESendEvent (int type, unsigned int mask, int kind, int arg, int context);
+
+int
+winHIMCtoContext (DWORD hIMC);
+
+void
+winCommitCompositionResult (int nContext, int nIndex, void *pData, int nLen);
+
+Bool
+winHIMCCompositionDraw(DWORD hIMC);
+
+/*
+ * winime.c
+ */
+
+Bool
+winInitImServer ();
+#endif
+
 /*
  * END DDX and DIX Function Prototypes
  */
Index: programs/Xserver/hw/xwin/winblock.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winblock.c,v
retrieving revision 1.5
diff -u -r1.5 winblock.c
--- programs/Xserver/hw/xwin/winblock.c	8 Jan 2005 13:01:03 -0000	1.5
+++ programs/Xserver/hw/xwin/winblock.c	16 Jan 2005 09:20:34 -0000
@@ -39,6 +39,7 @@
 extern HWND			g_hDlgDepthChange;
 extern HWND			g_hDlgExit;
 extern HWND			g_hDlgAbout;
+extern Bool			g_fIME;
 
 
 /* See Porting Layer Definition - p. 6 */
@@ -97,7 +98,7 @@
 	  && (g_hDlgAbout == 0
 	      || !IsDialogMessage (g_hDlgAbout, &msg)))
 	{
-	  DispatchMessage (&msg);
+	  winProcessMessage(&msg);
 	}
     }
 }
Index: programs/Xserver/hw/xwin/winconfig.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winconfig.c,v
retrieving revision 1.8
diff -u -r1.8 winconfig.c
--- programs/Xserver/hw/xwin/winconfig.c	8 Dec 2004 15:48:15 -0000	1.8
+++ programs/Xserver/hw/xwin/winconfig.c	16 Jan 2005 09:20:35 -0000
@@ -39,6 +39,16 @@
 #include "XKBsrv.h"
 #endif
 
+
+/*
+ * References to external symbols
+ */
+
+#ifdef XWIN_WINIME
+extern Bool			g_fIME;
+#endif
+
+
 #ifdef XWIN_XF86CONFIG
 #ifndef CONFIGPATH
 #define CONFIGPATH  "%A," "%R," \
@@ -333,7 +343,11 @@
 	   Same might apply for chinese, korean and other symbol languages
 	   too */
         layoutNum = (layoutNum & 0xffff);
-	if (keyboardType == 7)
+	if ((keyboardType == 7)
+#ifdef XWIN_WINIME
+	    && !g_fIME
+#endif
+	    )
 	  {
 	    /* Japanese layouts have problems with key event messages
 	       such as the lack of WM_KEYUP for Caps Lock key.
Index: programs/Xserver/hw/xwin/winglobals.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winglobals.c,v
retrieving revision 1.5
diff -u -r1.5 winglobals.c
--- programs/Xserver/hw/xwin/winglobals.c	22 Nov 2004 14:12:33 -0000	1.5
+++ programs/Xserver/hw/xwin/winglobals.c	16 Jan 2005 09:20:35 -0000
@@ -114,6 +114,15 @@
 #endif
 
 
+#ifdef XWIN_WINIME
+/*
+ * IME variables
+ */
+
+Bool			g_fIME = FALSE;
+#endif
+
+
 /*
  * Re-initialize global variables that are invalidated
  * by a server reset.
Index: programs/Xserver/hw/xwin/winmultiwindowwindow.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winmultiwindowwindow.c,v
retrieving revision 1.9
diff -u -r1.9 winmultiwindowwindow.c
--- programs/Xserver/hw/xwin/winmultiwindowwindow.c	12 Jan 2005 16:10:00 -0000	1.9
+++ programs/Xserver/hw/xwin/winmultiwindowwindow.c	16 Jan 2005 09:20:35 -0000
@@ -605,7 +605,7 @@
   HMODULE		hInstance;
   int			iReturn;
   char			pszClass[512];
-  
+
 #if CYGMULTIWINDOW_DEBUG
   ErrorF ("winDestroyWindowsWindow\n");
 #endif
@@ -631,7 +631,7 @@
     {
       if (g_hDlgDepthChange == 0 || !IsDialogMessage (g_hDlgDepthChange, &msg))
 	{
-	  DispatchMessage (&msg);
+	  winProcessMessage(&msg);
 	}
     }
 
Index: programs/Xserver/hw/xwin/winmultiwindowwndproc.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winmultiwindowwndproc.c,v
retrieving revision 1.4
diff -u -r1.4 winmultiwindowwndproc.c
--- programs/Xserver/hw/xwin/winmultiwindowwndproc.c	15 Sep 2004 17:58:50 -0000	1.4
+++ programs/Xserver/hw/xwin/winmultiwindowwndproc.c	16 Jan 2005 09:20:36 -0000
@@ -38,6 +38,12 @@
 #if CYGDEBUG
 #include "winmessages.h"
 #endif
+#ifdef XWIN_WINIME
+#define _WINIME_SERVER_
+#include "winime.h"
+#include "winimestr.h"
+#include <imm.h>
+#endif
 
 /*
  * External global variables
@@ -269,6 +275,39 @@
 
 
 /*
+ *
+ */
+
+char*
+WideToUTF8(int iUnicodeSize, wchar_t *pwszUnicodeStr, int *pLen)
+{
+  char			*pszUTF8 = NULL;
+
+  /* Convert to UTF8 */
+  *pLen = WideCharToMultiByte (CP_UTF8,
+			       0,
+			       (LPCWSTR)pwszUnicodeStr,
+			       iUnicodeSize/2,
+			       NULL,
+			       0,
+			       NULL,
+			       NULL);
+  pszUTF8 = (char *) malloc (*pLen+1);
+  WideCharToMultiByte (CP_UTF8,
+		       0,
+		       (LPCWSTR)pwszUnicodeStr,
+		       iUnicodeSize/2,
+		       pszUTF8,
+		       *pLen,
+		       NULL,
+		       NULL);
+  pszUTF8[*pLen] = '\0';
+
+  return pszUTF8;
+}
+
+
+/*
  * winTopLevelWindowProc - Window procedure for all top-level Windows windows.
  */
 
@@ -380,6 +419,11 @@
 	       WIN_WID_PROP,
 	       (HANDLE)winGetWindowID (((LPCREATESTRUCT) lParam)->lpCreateParams));
 
+#ifdef XWIN_WINIME
+      /* Disable IME by default */
+      ImmAssociateContext (hwnd, (HIMC) NULL);
+#endif
+
       /*
        * Make X windows' Z orders sync with Windows windows because
        * there can be AlwaysOnTop windows overlapped on the window
@@ -680,6 +724,13 @@
       ErrorF ("winTopLevelWindowProc - WM_*KEYDOWN\n");
 #endif
 
+#ifdef XWIN_WINIME
+      /*
+       * Ignore IME process key
+       */
+      if (wParam == VK_PROCESSKEY) return 0;
+#endif
+
       /*
        * Don't pass Alt-F4 key combo to root window,
        * let Windows translate to WM_CLOSE and close this top-level window.
@@ -744,6 +795,13 @@
       ErrorF ("winTopLevelWindowProc - WM_*KEYUP\n");
 #endif
 
+#ifdef XWIN_WINIME
+      /*
+       * Ignore IME process key
+       */
+      if (wParam == VK_PROCESSKEY) return 0;
+#endif
+
       /* Pass the message to the root window */
       SendMessage (hwndScreen, message, wParam, lParam);
       return 0;
@@ -1051,6 +1109,16 @@
 	}
       break;
 
+#ifdef XWIN_WINIME
+    case WM_IME_NOTIFY:
+    case WM_IME_STARTCOMPOSITION:
+    case WM_IME_COMPOSITION:
+    case WM_IME_ENDCOMPOSITION:
+    case WM_IME_CHAR:
+    case WM_CHAR:
+      return winIMEMessageHandler (hwnd, message, wParam, lParam);
+#endif
+
     default:
       break;
     }
Index: programs/Xserver/hw/xwin/winprocarg.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winprocarg.c,v
retrieving revision 1.10
diff -u -r1.10 winprocarg.c
--- programs/Xserver/hw/xwin/winprocarg.c	15 Dec 2004 12:22:39 -0000	1.10
+++ programs/Xserver/hw/xwin/winprocarg.c	16 Jan 2005 09:20:36 -0000
@@ -54,6 +54,9 @@
 extern Bool			g_fNoHelpMessageBox;                     
 extern Bool			g_fSoftwareCursor;
 extern Bool			g_fSilentDupError;
+#ifdef XWIN_WINIME
+extern Bool			g_fIME;
+#endif
 
 /* globals required by callback function for monitor information */
 struct GetMonitorInfoData {
@@ -1259,6 +1262,17 @@
       return 1;
     }
 
+#ifdef XWIN_WINIME
+  /*
+   * Look for the '-ime' argument
+   */
+  if (IS_OPTION ("-ime"))
+    {
+      g_fIME = TRUE;
+      return 1;
+    }
+#endif
+
   /*
    * Look for the '-fp' argument
    */
Index: programs/Xserver/hw/xwin/winscrinit.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winscrinit.c,v
retrieving revision 1.8
diff -u -r1.8 winscrinit.c
--- programs/Xserver/hw/xwin/winscrinit.c	15 Nov 2004 15:06:48 -0000	1.8
+++ programs/Xserver/hw/xwin/winscrinit.c	16 Jan 2005 09:20:36 -0000
@@ -75,6 +75,9 @@
 extern miPointerScreenFuncRec	g_winPointerCursorFuncs;
 extern int			g_iScreenPrivateIndex;
 extern Bool                     g_fSoftwareCursor;
+#ifdef XWIN_WINIME
+extern Bool			g_fIME;
+#endif
 
 
 /*
@@ -467,6 +470,20 @@
     }
 #endif
 
+#ifdef XWIN_WINIME
+  if (g_fIME)
+    {
+      winWinIMEExtensionInit ();
+
+      /* Initialize XIM server */
+      if (!winInitImServer())
+	{
+	  ErrorF ("winFinishScreenInitFB - winInitImServer () failed.\n");
+	  return FALSE;
+	}
+    }
+#endif
+
   /* Handle rootless mode */
   if (pScreenInfo->fRootless)
     {
Index: programs/Xserver/hw/xwin/winwakeup.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winwakeup.c,v
retrieving revision 1.4
diff -u -r1.4 winwakeup.c
--- programs/Xserver/hw/xwin/winwakeup.c	15 Nov 2004 15:06:48 -0000	1.4
+++ programs/Xserver/hw/xwin/winwakeup.c	16 Jan 2005 09:20:36 -0000
@@ -63,7 +63,7 @@
 	  && (g_hDlgAbout == 0
 	      || !IsDialogMessage (g_hDlgAbout, &msg)))
 	{
-	  DispatchMessage (&msg);
+	  winProcessMessage(&msg);
 	}
     }
 }
Index: programs/Xserver/hw/xwin/winwin32rootlesswndproc.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winwin32rootlesswndproc.c,v
retrieving revision 1.5
diff -u -r1.5 winwin32rootlesswndproc.c
--- programs/Xserver/hw/xwin/winwin32rootlesswndproc.c	4 Nov 2004 11:52:22 -0000	1.5
+++ programs/Xserver/hw/xwin/winwin32rootlesswndproc.c	16 Jan 2005 09:20:37 -0000
@@ -40,6 +40,12 @@
 #if CYGDEBUG
 #include "winmessages.h"
 #endif
+#ifdef XWIN_WINIME
+#define _WINIME_SERVER_
+#include "winime.h"
+#include "winimestr.h"
+#include <imm.h>
+#endif
 
 
 /*
@@ -469,6 +475,11 @@
       SetProp (hwnd,
 	       WIN_WINDOW_PROP,
 	       (HANDLE)((LPCREATESTRUCT) lParam)->lpCreateParams);
+
+#ifdef XWIN_WINIME
+      /* Disable IME by default */
+      ImmAssociateContext (hwnd, (HIMC) NULL);
+#endif
       return 0;
 
     case WM_CLOSE:
@@ -748,6 +759,13 @@
       winDebug ("winMWExtWMWindowProc - WM_*KEYDOWN\n");
 #endif
 
+#ifdef XWIN_WINIME
+      /*
+       * Ignore IME process key
+       */
+      if (wParam == VK_PROCESSKEY) return 0;
+#endif
+
       /*
        * Don't pass Alt-F4 key combo to root window,
        * let Windows translate to WM_CLOSE and close this top-level window.
@@ -776,6 +794,13 @@
       winDebug ("winMWExtWMWindowProc - WM_*KEYUP\n");
 #endif
 
+#ifdef XWIN_WINIME
+      /*
+       * Ignore IME process key
+       */
+      if (wParam == VK_PROCESSKEY) return 0;
+#endif
+
       /* Pass the message to the root window */
       SendMessage (hwndScreen, message, wParam, lParam);
       return 0;
@@ -1309,6 +1334,16 @@
       ErrorF ("winMWExtWMWindowProc - WM_UNMANAGE\n");
       break;
 
+#ifdef XWIN_WINIME
+    case WM_IME_NOTIFY:
+    case WM_IME_STARTCOMPOSITION:
+    case WM_IME_COMPOSITION:
+    case WM_IME_ENDCOMPOSITION:
+    case WM_IME_CHAR:
+    case WM_CHAR:
+      return winIMEMessageHandler (hwnd, message, wParam, lParam);
+#endif
+
     default:
       break;
     }
Index: programs/Xserver/hw/xwin/winwindowswm.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winwindowswm.c,v
retrieving revision 1.2
diff -u -r1.2 winwindowswm.c
--- programs/Xserver/hw/xwin/winwindowswm.c	21 Jun 2004 13:19:32 -0000	1.2
+++ programs/Xserver/hw/xwin/winwindowswm.c	16 Jan 2005 09:20:37 -0000
@@ -240,7 +240,7 @@
        */
       if (!pHead)
 	{
-	  pHead = (WMEventPtr *) xalloc (sizeof (WMEventPtr));
+	  pHead = (WMEventPtr *) xalloc (sizeof (WMEventRec));
 	  if (!pHead ||
 	      !AddResource (eventResource, EventType, (pointer)pHead))
 	    {
Index: programs/Xserver/hw/xwin/winwndproc.c
===================================================================
RCS file: /cvs/xorg/xc/programs/Xserver/hw/xwin/winwndproc.c,v
retrieving revision 1.9
diff -u -r1.9 winwndproc.c
--- programs/Xserver/hw/xwin/winwndproc.c	10 Jan 2005 13:13:08 -0000	1.9
+++ programs/Xserver/hw/xwin/winwndproc.c	16 Jan 2005 09:20:37 -0000
@@ -58,6 +58,7 @@
 extern HWND			g_hwndKeyboardFocus;
 extern Bool			g_fSoftwareCursor;
 extern DWORD			g_dwCurrentThreadID;
+extern Bool			g_fIME;
 
 /*
  * Called by winWakeupHandler
@@ -1257,3 +1258,29 @@
 
   return DefWindowProc (hwnd, message, wParam, lParam);
 }
+
+void
+winProcessMessage (LPMSG lpMsg)
+{
+#if 0
+  if (lpMsg->message == WM_KEYUP || lpMsg->message == WM_KEYDOWN)
+    {
+      winDebug("winProcessMessage %d\n", GetTickCount());
+      winDebug("msg:0x%02x wParam:0x%04x time:%d", lpMsg->message, lpMsg->wParam, lpMsg->time);
+      winDebug("lParam:count(0x%04x) scancode(0x%02x) flag(0x%x,0x%x,0x%x,0x%x)\n",
+	       LOWORD (lpMsg->lParam),
+	       LOBYTE (HIWORD (lpMsg->lParam)),
+	       (HIBYTE (HIWORD (lpMsg->lParam)) & 0x01) >> 0,
+	       (HIBYTE (HIWORD (lpMsg->lParam)) & 0x20) >> 5,
+	       (HIBYTE (HIWORD (lpMsg->lParam)) & 0x40) >> 6,
+	       (HIBYTE (HIWORD (lpMsg->lParam)) & 0x80) >> 7);
+    }
+#endif
+#ifdef XWIN_WINIME
+  if (g_fIME)
+    {
+      TranslateMessage(lpMsg);
+    }
+#endif
+  DispatchMessage (lpMsg);
+}
